### DISCUSSION TITLE
support for `textDocument/diagnostic`

### DISCUSSION THREAD

**Discussion by waymondo** [joaotavora/eglot#1290] [Q&A] [answered]
I'm not sure how common this is across other LSP servers, but `ruby-lsp` does not use `textDocument/publishDiagnostics` to publish diagnostics - it only uses that event to clear them out on document close.

Instead, their VSCode plugin relies on polling for diagnostics in an interval with `textDocument/diagnostic`. a quick look through the Eglot source and I don't see evidence of a similar mechanism. is this correct? in the `ruby-lsp` wiki it seems like the Neovim LSP client also doesn't support this and so there is [some additional code](https://github.com/Shopify/ruby-lsp/blob/main/EDITORS.md#neovim-lsp) someone has created for this purpose.

if this currently isn't a feature of Eglot, would it make sense to entertain as a new feature? are there any other known LSP servers that rely on polling for diagnostics like this?

**Comment by joaotavora** [comment]
> . is this correct? 

Yes.  This is called "pull diagnostics" or something.

> if this currently isn't a feature of Eglot, would it make sense to entertain as a new feature?

Yes, it would.  In fact, this should be closer to Emacs's model of getting diagnostics for other languages.  **However** this is an very tricky part of Eglot and I would like to not break it or over-complexify it.  So a carefully tested and well-designed implementation must be done.

Please send email to bug-gnu-emacs@gnu.org to open a bug report with this feature request, include these messages as context, and remember to CC me.  Thanks.

**Reply by waymondo** [reply]
thanks! will do. what's your email address for CC-ing?

**Reply by joaotavora** [reply]
Find it in the Eglot source

**Comment by WardBrian** [comment]
I can't find any mail on the gnu list about this (though that could easily be a failure on my part) -- did this ever receive further attention?

**Reply by joaotavora** [reply, marked as answer]
There is now limited support for this server capability in the Emacs  master branch.  You can try to compile Emacs from the `master` branch, or wait some time to grab Eglot from GNU-devel ELPA (https://elpa.gnu.org/devel/eglot.html).  However not that this support is limited, chiefly for lack of capable servers.  From `etc/EGLOT-NEWS`:

```
** Support for pull diagnostics (github#1559, github#1290)

For servers supporting the 'diagnosticProvider' capability, Eglot
requests diagnostics on-demand rather than relying solely on
server-pushed 'publishDiagnostics' notifications.  This may provide a
more responsive user experience.  The 'ty' server is known to support it
while the 'tsgo' server is known to support is exclusively.  However, no
server has been found to support the 'relatedDocumentSupport'
sub-capability, which e.g. introducing a problem in one file to bring
diagnostics for another related file.  The traditional push diagnostics,
as supported by some servers, are at the moment still a superior user
experience in this particular regard.
```
